### PR TITLE
net/reconnect: introduce StreamSession to aid writting asynchronous login

### PR DESCRIPTION
**Pull request by amery**

### CODE REVIEW

File: net/reconnect/stream.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+	s.wg.OnError(s.OnError)
+	s.wg.SetDefaults()
+
+	s.in = make(chan T1)
+	s.out = make(chan T2, s.QueueSize)
+	return nil
+}
+
+// revive:disable:cognitive-complexity
+func (s *StreamSession[T1, T2]) setDefaults() error {
```
[the remark this answers is not in the document]
Comment (amery) [reply net/reconnect/stream.go:85]: in that case how do you make it unbuffered?

**Comment by amery** [comment]
amended to use `Input` and `Output` instead of `T1` and `T2`

moved the panic into a `mustStarted()` function

**Review by karasz — APPROVED** [review]
LGTM

**Review by karasz — COMMENTED** [review]
+1

**Review by karasz — APPROVED** [review]

**Pull request merged**